Add unit tests for Meter Reading DocType

Levinmunyelele opened the pull request:
This PR adds unit tests for the Meter Reading DocType to ensure functionality and data integrity. Covers basic validations and business logic.

Levinmunyelele committed:
feat: Add unit tests for Meter Reading DocType

Levinmunyelele committed:
Add unit tests for Meter Reading DocType

muruthigitau reviewed (commented):
> **Hey @Levinmunyelele,**
>
> I had a look at the Meter Reading tests and wanted to share some feedback and suggestions. Overall, things look pretty solid, but there are a few spots where we could tidy up the code and improve readability.
>
> Happy to discuss or help out with any changes whenever you want!

## 🧩 General Observations

* ✅ **Structure**: The test suite is generally well-organized. Use of `setUpClass` for shared test setup is appropriate.
* ✅ **Testing Approach**: A good balance between unit and integration tests is observed.
* ✅ **Documentation**: Docstrings are present and helpful, providing clear insight into the test intent.

---

## Code Issues and Fixes

### 1. **Syntax Error, Misplaced Logic, `serial_no` Handling & Missing `customer` Definition**

```python
def create_warranty_claim(...):
    ...
    if not frappe.db.exists("Warranty Claim", ...):
        create_warranty_claim(...)
```

**Issues**:

* ❌ **Syntax Error**: Misplaced logic and conditional calls inside the function leads to infinite recursion and unreachable code.
* ❌ **`serial_no` Handling**: `None` or empty string values aren't handled, which may cause issues with validation or database constraints.
* ❌ **Missing `customer` Definition**: The `customer` variable is referenced but never initialized or passed into the function.

✅ **Fixes**:

* Move the `create_warranty_claim(...)` calls *outside* the function to avoid logical/syntax errors.
* Safely handle `serial_no` with:

```python
"serial_no": serial_no or ""
# or
if serial_no:
    doc.serial_no = serial_no
```

* Define a `customer` before using it:


---

### 2. **Manual Setting of `name` Field**

```python
"name": "WC-001-TEST-MULTIPLE"
```

**Issue**:
Manually setting `name` risks collisions since it's usually system-generated.

✅ **Fix**:
Avoid this unless:

* You're using `set_name=True`
* You ensure it's unique

---

### 3. ⚠️ **Use of `ignore_links=True`**

```python
.insert(ignore_permissions=True, ignore_links=True)
```

**Issue**:
Suppressing link validation may lead to inconsistent or invalid references in data.

✅ **Fix**:
Only use this in isolated tests. Prefer:

```python
.insert(ignore_permissions=True)
```

---

### 4. **Improper Multi-line `serial_no` Usage**

```python
"serial_no": "SN001\nSN002\nSN004"
```

**Issue**:
`serial_no` is likely a single Link field. Multi-line strings won’t work.

✅ **Fix**:
Use only a single valid serial number per warranty claim.

---
### 5. ⚠️ **Missing Test Cleanup (`tearDownClass`)**

**Issue**:
Without cleanup, tests leave behind residual data, which may cause interference between test runs.

✅ **Fix**:
Add cleanup logic

Levinmunyelele committed:
Fix: meter reading tests - handled serial_no issues, added cleanup, removed ignore_links

Levinmunyelele committed:
Fix warranty claim tests: serial_no handling, company field, cleanup

Levinmunyelele committed:
Move test_meter_reading.py back to correct path

Levinmunyelele commented:
Hi @muruthigitau, I've implemented all your feedback:
- Fixed serial_no handling and company field
- Removed multi-line serials
- Added tearDownClass cleanup
- Moved test_meter_reading.py back to the correct path

Kindly review when you get a chance. Thank you

muruthigitau commented on utility_billing/hooks.py:
NOTE: 'crm' should not be removed from required_apps as it is used in the app

Levinmunyelele committed:
fix: add 'crm' to required_apps in hooks.py

Levinmunyelele commented:
All updates done. Let me know if any further changes are needed. Thanks again.
